---
title: Правило анализа кода CA2153 для исключений поврежденного состояния (анализ кода)
description: Сведения о правиле анализа кода правила анализа кода CA2153 для исключений поврежденного состояния
ms.date: 02/19/2019
ms.topic: reference
author: gewarren
ms.author: gewarren
ms.openlocfilehash: 32d9c6ef19d0cf1fa869f21cff7dc90d95c44277
ms.sourcegitcommit: ddf7edb67715a5b9a45e3dd44536dabc153c1de0
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 02/06/2021
ms.locfileid: "99714876"
---
# <a name="ca2153-avoid-handling-corrupted-state-exceptions"></a><span data-ttu-id="5f3a9-103">CA2153. Не обрабатывайте исключения поврежденного состояния</span><span class="sxs-lookup"><span data-stu-id="5f3a9-103">CA2153: Avoid handling Corrupted State Exceptions</span></span>

| | <span data-ttu-id="5f3a9-104">Значение</span><span class="sxs-lookup"><span data-stu-id="5f3a9-104">Value</span></span> |
|-|-|
| <span data-ttu-id="5f3a9-105">**Идентификатор правила**</span><span class="sxs-lookup"><span data-stu-id="5f3a9-105">**Rule ID**</span></span> |<span data-ttu-id="5f3a9-106">CA2153</span><span class="sxs-lookup"><span data-stu-id="5f3a9-106">CA2153</span></span>|
| <span data-ttu-id="5f3a9-107">**Категория**</span><span class="sxs-lookup"><span data-stu-id="5f3a9-107">**Category**</span></span> |[<span data-ttu-id="5f3a9-108">Безопасность</span><span class="sxs-lookup"><span data-stu-id="5f3a9-108">Security</span></span>](security-warnings.md)|
| <span data-ttu-id="5f3a9-109">**Исправление не критическое или не критическое**</span><span class="sxs-lookup"><span data-stu-id="5f3a9-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="5f3a9-110">Не критическое</span><span class="sxs-lookup"><span data-stu-id="5f3a9-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="5f3a9-111">Причина</span><span class="sxs-lookup"><span data-stu-id="5f3a9-111">Cause</span></span>

<span data-ttu-id="5f3a9-112">[Исключения поврежденного состояния (CSE)](/archive/msdn-magazine/2009/february/clr-inside-out-handling-corrupted-state-exceptions) указывают на то, что в процессе существует повреждение памяти.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-112">[Corrupted State Exceptions (CSEs)](/archive/msdn-magazine/2009/february/clr-inside-out-handling-corrupted-state-exceptions) indicate that memory corruption exists in your process.</span></span> <span data-ttu-id="5f3a9-113">Если перехватывать их вместо того, чтобы позволить процессу завершиться сбоем, это может привести к уязвимостям в системе безопасности, если злоумышленнику удастся поместить эксплойт в поврежденную область памяти.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-113">Catching these rather than allowing the process to crash can lead to security vulnerabilities if an attacker can place an exploit into the corrupted memory region.</span></span>

## <a name="rule-description"></a><span data-ttu-id="5f3a9-114">Описание правила</span><span class="sxs-lookup"><span data-stu-id="5f3a9-114">Rule description</span></span>

<span data-ttu-id="5f3a9-115">Исключение CSE указывает на то, что состояние процесса было повреждено, и не перехватывается системой.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-115">CSE indicates that the state of a process has been corrupted and not caught by the system.</span></span> <span data-ttu-id="5f3a9-116">В сценарии с поврежденным состоянием общий обработчик перехватывает только исключение, если вы помечаете метод <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute?displayProperty=fullName> атрибутом.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-116">In the corrupted state scenario, a general handler only catches the exception if you mark your method with the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute?displayProperty=fullName> attribute.</span></span> <span data-ttu-id="5f3a9-117">По умолчанию [Среда](../../../standard/clr.md) CLR не вызывает обработчики catch для CSE.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-117">By default, the [Common Language Runtime (CLR)](../../../standard/clr.md) does not invoke catch handlers for CSEs.</span></span>

<span data-ttu-id="5f3a9-118">Наиболее надежный вариант — Разрешить аварийное завершение процесса без перехвата исключений такого рода.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-118">The safest option is to allow the process to crash without catching these kinds of exceptions.</span></span> <span data-ttu-id="5f3a9-119">Даже код регистрации может позволить злоумышленникам использовать ошибки повреждения памяти.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-119">Even logging code can allow attackers to exploit memory corruption bugs.</span></span>

<span data-ttu-id="5f3a9-120">Это предупреждение активируется при перехвате CSE с помощью общего обработчика, который перехватывает все исключения, например `catch (System.Exception e)` или `catch` без параметра исключения.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-120">This warning triggers when catching CSEs with a general handler that catches all exceptions, for example, `catch (System.Exception e)` or `catch` with no exception parameter.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="5f3a9-121">Устранение нарушений</span><span class="sxs-lookup"><span data-stu-id="5f3a9-121">How to fix violations</span></span>

<span data-ttu-id="5f3a9-122">Чтобы устранить это предупреждение, выполните одно из следующих действий.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-122">To resolve this warning, do one of the following:</span></span>

- <span data-ttu-id="5f3a9-123">Удалите атрибут <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-123">Remove the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> attribute.</span></span> <span data-ttu-id="5f3a9-124">Это возвращается к поведению во время выполнения по умолчанию, когда CSE не передаются в обработчики catch.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-124">This reverts to the default run-time behavior where CSEs are not passed to catch handlers.</span></span>

- <span data-ttu-id="5f3a9-125">Удалите общий обработчик catch и используйте обработчики, перехватывающие исключения определенных типов.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-125">Remove the general catch handler in preference of handlers that catch specific exception types.</span></span> <span data-ttu-id="5f3a9-126">Это может включать CSE, предполагая, что код обработчика может безопасно их обменять (редкие).</span><span class="sxs-lookup"><span data-stu-id="5f3a9-126">This may include CSEs, assuming the handler code can safely handle them (rare).</span></span>

- <span data-ttu-id="5f3a9-127">Повторно вызовите CSE в обработчике Catch, который передает исключение вызывающему объекту и должен привести к завершению выполняющегося процесса.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-127">Rethrow the CSE in the catch handler, which passes the exception to the caller and should result in ending the running process.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="5f3a9-128">Когда следует подавлять предупреждения</span><span class="sxs-lookup"><span data-stu-id="5f3a9-128">When to suppress warnings</span></span>

<span data-ttu-id="5f3a9-129">Для этого правила отключать вывод предупреждений не следует.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-129">Do not suppress a warning from this rule.</span></span>

## <a name="pseudo-code-example"></a><span data-ttu-id="5f3a9-130">Пример псевдо-кода</span><span class="sxs-lookup"><span data-stu-id="5f3a9-130">Pseudo-code example</span></span>

### <a name="violation"></a><span data-ttu-id="5f3a9-131">Нарушение</span><span class="sxs-lookup"><span data-stu-id="5f3a9-131">Violation</span></span>

<span data-ttu-id="5f3a9-132">В приведенном ниже псевдокоде показан шаблон, обнаруживаемый этим правилом.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-132">The following pseudo-code illustrates the pattern detected by this rule.</span></span>

```csharp
[HandleProcessCorruptedStateExceptions]
// Method that handles CSE exceptions.
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Handle exception.
    }
}
```

### <a name="solution-1---remove-the-attribute"></a><span data-ttu-id="5f3a9-133">Решение 1. Удаление атрибута</span><span class="sxs-lookup"><span data-stu-id="5f3a9-133">Solution 1 - remove the attribute</span></span>

<span data-ttu-id="5f3a9-134">Удаление <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> атрибута гарантирует, что исключения поврежденного состояния не будут обрабатываться методом.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-134">Removing the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> attribute ensures that Corrupted State Exceptions are not handled by your method.</span></span>

```csharp
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Handle exception.
    }
}
```

### <a name="solution-2---catch-specific-exceptions"></a><span data-ttu-id="5f3a9-135">Решение 2. перехват особых исключений</span><span class="sxs-lookup"><span data-stu-id="5f3a9-135">Solution 2 - catch specific exceptions</span></span>

<span data-ttu-id="5f3a9-136">Удалите общий обработчик catch и перехватывайте только исключения определенных типов.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-136">Remove the general catch handler and catch only specific exception types.</span></span>

```csharp
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (IOException e)
    {
        // Handle IOException.
    }
    catch (UnauthorizedAccessException e)
    {
        // Handle UnauthorizedAccessException.
    }
}
```

### <a name="solution-3---rethrow"></a><span data-ttu-id="5f3a9-137">Решение 3. Повторное создание</span><span class="sxs-lookup"><span data-stu-id="5f3a9-137">Solution 3 - rethrow</span></span>

<span data-ttu-id="5f3a9-138">Вызовите исключение заново.</span><span class="sxs-lookup"><span data-stu-id="5f3a9-138">Rethrow the exception.</span></span>

```csharp
[HandleProcessCorruptedStateExceptions]
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Rethrow the exception.
        throw;
    }
}
```
